// ibm-watson/auth/index.js
"use strict";
/**
 * (C) Copyright IBM Corp. 2019.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
var ibm_cloud_sdk_core_1 = require("ibm-cloud-sdk-core");
exports.NoAuthAuthenticator = ibm_cloud_sdk_core_1.NoAuthAuthenticator;
exports.BasicAuthenticator = ibm_cloud_sdk_core_1.BasicAuthenticator;
exports.BearerTokenAuthenticator = ibm_cloud_sdk_core_1.BearerTokenAuthenticator;
exports.IamAuthenticator = ibm_cloud_sdk_core_1.IamAuthenticator;
exports.CloudPakForDataAuthenticator = ibm_cloud_sdk_core_1.CloudPakForDataAuthenticator;
exports.IamTokenManager = ibm_cloud_sdk_core_1.IamTokenManager;
exports.Cp4dTokenManager = ibm_cloud_sdk_core_1.Cp4dTokenManager;
exports.getAuthenticatorFromEnvironment = ibm_cloud_sdk_core_1.getAuthenticatorFromEnvironment;
